src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -33,6 +32,42 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
             return result;
         }
 
+        private static bool CanLowerToStringConcatenation(BoundInterpolatedString node)
+        {
+            foreach (var part in node.Parts)
+            {
+                if (part is BoundStringInsert fillin)
+                {
+                    // this is one of the expression holes
+                    if (fillin.HasErrors ||
+                        fillin.Value.Type?.SpecialType != SpecialType.System_String ||
+                        fillin.Alignment != null ||
+                        fillin.Format != null)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static string Unescape(string s)
```

Comment:
I suspect this method can be re-used in `VisitInterpolatedString` (`LocalRewriter_StringInterpolation.cs` line 78)

Reply:
No it cannot. `String.Format` requires `'{'` to be escaped as `"{{"` and `'{'` to be escaped as `"}}"`.

Reply:
I'm sorry, I don't see the difference between this method and the code I referred to.

```C#
                var builder = PooledStringBuilder.GetInstance();
                ...
                int formatLength = formatText.Length;
                for (int i = 0; i < formatLength; i++)
                {
                    char c = formatText[i];
                    builder.Builder.Append(c);
                    if ((c == '{' || c == '}') && (i + 1) < formatLength && formatText[i + 1] == c)
                    {
                        i++;
                    }
                }
                ... builder.ToStringAndFree() ...
```

---
In reply to: [190505462](https://github.com/dotnet/roslyn/pull/26612#discussion_r190505462) [](ancestors = 190505462)

Reply:
I'm confused. What code are you referring too?

Or are you saying that line 86:

     formatString.Builder.Append(part.ConstantValue.StringValue);

does the same thing as:

            int formatLength = s.Length;
            for (int i = 0; i < formatLength; i++)
            {
                char c = s[i];
                stringBuilder.Append(c);
                if ((c == '{' || c == '}') && (i + 1) < formatLength && s[i + 1] == c)
                {
                    i++;
                }
            }

?

It's not the same. When lowering an interpolation to ´string.Format`, any `"{{"` must remain `"{{"`, because ´string.Format` needs them to be escaped. When lowering to string concatenation, they need to be unescaped to `"{"`.

Reply:
My bad, I was looking at a local copy of the file before your change. Sorry for the confusion.

---
In reply to: [191042562](https://github.com/dotnet/roslyn/pull/26612#discussion_r191042562) [](ancestors = 191042562)